clearnode/channel_service.go

```diff
@@ -0,0 +1,263 @@
+	if !ok {
+		return CloseChannelResponse{}, fmt.Errorf("invalid signature")
+	}
+
+	asset, err := GetAssetByToken(s.db, channel.Token, channel.ChainID)
+	if err != nil {
+		logger.Error("failed to find asset", "error", err)
+		return CloseChannelResponse{}, fmt.Errorf("failed to find asset for token %s on chain %d", channel.Token, channel.ChainID)
+	}
+
+	ledger := GetWalletLedger(s.db, channel.Wallet)
+	balance, err := ledger.Balance(channel.Wallet, asset.Symbol)
+	if err != nil {
+		logger.Error("failed to check participant balance", "error", err)
+		return CloseChannelResponse{}, fmt.Errorf("failed to check participant balance")
+	}
+	if balance.IsNegative() {
+		logger.Error("negative balance", "balance", balance.String())
+		return CloseChannelResponse{}, fmt.Errorf("negative balance")
+	}
+
+	rawBalance := balance.Shift(int32(asset.Decimals)).BigInt()
+	channelAmount := new(big.Int).SetUint64(channel.Amount)
+	if channelAmount.Cmp(rawBalance) < 0 {
+		return CloseChannelResponse{}, fmt.Errorf("resize this channel first")
+	}
+
+	allocations := []nitrolite.Allocation{
+		{
+			Destination: common.HexToAddress(params.FundsDestination),
+			Token:       common.HexToAddress(channel.Token),
+			Amount:      rawBalance,
+		},
+		{
+			Destination: s.signer.GetAddress(),
+			Token:       common.HexToAddress(channel.Token),
+			Amount:      new(big.Int).Sub(channelAmount, rawBalance),
+		},
+	}
+
+	stateDataHex := "0x"
+	stateDataBytes, err := hexutil.Decode(stateDataHex)
+	if err != nil {
+		logger.Error("failed to decode state data hex", "error", err)
+		return CloseChannelResponse{}, fmt.Errorf("failed to decode state data hex")
+	}
+	encodedState, err := nitrolite.EncodeState(common.HexToHash(channel.ChannelID), nitrolite.IntentFINALIZE, big.NewInt(int64(channel.Version)+1), stateDataBytes, allocations)
+	if err != nil {
+		logger.Error("failed to encode state hash", "error", err)
+		return CloseChannelResponse{}, fmt.Errorf("failed to encode state hash")
+	}
+	stateHash := crypto.Keccak256Hash(encodedState).Hex()
+	sig, err := s.signer.NitroSign(encodedState)
+	if err != nil {
+		logger.Error("failed to sign state", "error", err)
+		return CloseChannelResponse{}, fmt.Errorf("failed to sign state")
+	}
+
+	resp := CloseChannelResponse{
+		ChannelID: channel.ChannelID,
+		Intent:    uint8(nitrolite.IntentFINALIZE),
+		Version:   channel.Version + 1,
+		StateData: stateDataHex,
+		StateHash: stateHash,
+		Signature: Signature{
+			V: sig.V,
+			R: hexutil.Encode(sig.R[:]),
+			S: hexutil.Encode(sig.S[:]),
+		},
+	}
+
+	for _, alloc := range allocations {
+		resp.FinalAllocations = append(resp.FinalAllocations, Allocation{
+			Participant:  alloc.Destination.Hex(),
+			TokenAddress: alloc.Token.Hex(),
+			Amount:       alloc.Amount,
+		})
+	}
+
+	return resp, nil
+}
+
+// checkChallengedChannels checks if the participant has any channels in the challenged state.
+func checkChallengedChannels(tx *gorm.DB, wallet string) error {
```

Comment:
If it won't be used anywhere except ChannelService, I'd suggest declaring it as its method